packages/cqrs/lib/src/cqrs_result.dart

```diff
@@ -0,0 +1,202 @@
+  bool get isInvalid => error == CqrsError.validation;
+}
+
+/// Generic result for CQRS query result. Can be either [CqrsQuerySuccess]
+/// or [CqrsQueryFailure].
+sealed class CqrsQueryResult<T, E> extends Equatable {
+  /// Creates a [CqrsQueryResult] class.
+  const CqrsQueryResult();
+
+  /// Casts to [CqrsQuerySuccess] or `null`.
+  CqrsQuerySuccess<T, E>? get asSuccess => switch (this) {
+        final CqrsQuerySuccess<T, E> s => s,
+        CqrsQueryFailure() => null,
+      };
+
+  /// Casts to [CqrsQueryFailure] or `null`.
+  CqrsQueryFailure<T, E>? get asFailure => switch (this) {
+        CqrsQuerySuccess() => null,
+        final CqrsQueryFailure<T, E> f => f,
+      };
+
+  /// Whether this instance can be casted to [CqrsQuerySuccess].
+  bool get isSuccess => asSuccess != null;
+
+  /// Whether this instance can be casted to [CqrsQueryFailure].
+  bool get isFailure => asFailure != null;
+
+  /// Returns [data] of type [T] for [CqrsQuerySuccess] and `null`
+  /// for [CqrsQueryFailure].
+  T? get data => asSuccess?.data;
+
+  /// Returns [error] of type [E] for [CqrsQueryFailure] and `null`
+  /// for [CqrsQuerySuccess].
+  E? get error => asFailure?.error;
+}
+
+/// Generic class which represents a result of succesful query execution.
+final class CqrsQuerySuccess<T, E> extends CqrsQueryResult<T, E> {
+  /// Creates a [CqrsQuerySuccess] class.
+  const CqrsQuerySuccess(this.data);
+
+  @override
+  final T data;
+
+  @override
+  List<Object?> get props => [data];
+}
+
+/// Generic class which represents a result of unsuccesful query execution.
+final class CqrsQueryFailure<T, E> extends CqrsQueryResult<T, E> {
+  /// Creates a [CqrsQueryFailure] class.
+  const CqrsQueryFailure(this.error);
+
+  @override
+  final E error;
+
+  @override
+  List<Object?> get props => [error];
+}
+
+/// Generic result for CQRS command result. Can be either [CqrsCommandSuccess]
+/// or [CqrsCommandFailure].
+sealed class CqrsCommandResult<E> extends Equatable {
+  /// Creates a [CqrsCommandResult] class.
+  const CqrsCommandResult();
+
+  /// Casts to [CqrsCommandSuccess] or `null`.
+  CqrsCommandSuccess<E>? get asSuccess => switch (this) {
+        final CqrsCommandSuccess<E> s => s,
+        CqrsCommandFailure() => null,
+      };
+
+  /// Casts to [CqrsCommandFailure] or `null`.
+  CqrsCommandFailure<E>? get asFailure => switch (this) {
+        CqrsCommandSuccess() => null,
+        final CqrsCommandFailure<E> f => f,
+      };
+
+  /// Whether this instance can be casted to [CqrsCommandSuccess].
+  bool get isSuccess => asSuccess != null;
+
+  /// Whether this instance can be casted to [CqrsCommandFailure].
+  bool get isFailure => asFailure != null;
+
+  /// Returns an [error] of type [E] for [CqrsCommandFailure] and `null`
+  /// for [CqrsCommandSuccess].
+  E? get error => asFailure?.error;
+
+  /// Returns a list of [ValidationError] errors.
+  List<ValidationError> get validationErrors;
```

Comment:
Any particular reason why we are not defining it only on the error variant?

Reply:
I would say convenience. For command result it might be more handy to write:

```dart
if (result.isInvalid) {
  handleValidationErrors(result.validation)
}
```

than having to exhaust all posibble result types as so:

```dart
switch (result) {
  CqrsCommandFailure(
    isInvalid: true,
    :final validationErrors,
  ) =>
    handleValidationErrors(validationErrors),
  _ => null,
};
```

Reply:
In my recent changes I changed that only `CFailure` will have `validationErrors` as field.